### ISSUE TITLE
It seems that the weights in the file named 'distilled_inv_64.pth' are incorrect

### ISSUE THREAD

**Issue by LockFaiz** [ViviaHu/STA#1]
I believe"inverse" or "inver" operation refers to reciprocal, which expressed by y=(1/x)

However, upon attempting to reproduce the "inver" operation, I found that it appears to be different.
![image](https://github.com/ViviaHu/STA/assets/70676003/f1c7f0b2-d7a8-407c-ac84-24cc496d0047)

**Comment by LockFaiz** [comment]
So inverse network in this work is trained to fit (1/x) for x > 2, it does not matter for 0<x <2

**Comment by Ewedihalehu** [comment]
> 因此，这项工作中的逆网络被训练为适合 x > 2 的 （1/x），而 0<x <2 无关紧要

请问，你跑出来的精度和论文里相同吗，我怎么跑出来差了好多呀，是没有微调的原因吗

**Comment by LockFaiz** [comment]
不需要微调吧，我们组应该是有人复现了。

**Comment by LockFaiz** [comment]
不需要微调吧，我们组应该是有人复现了。

**Comment by condy0919** [comment]
> > 因此，这项工作中的逆网络被训练为适合 x > 2 的 （1/x），而 0<x <2 无关紧要
> 
> 请问，你跑出来的精度和论文里相同吗，我怎么跑出来差了好多呀，是没有微调的原因吗

复现还是没问题的，我尝试着把它这部分给整到 llama2-7B 模型上做情感分类效果还是不错的

**Comment by condy0919** [comment]
😭 我跑了一下，发现我复现不出来 @Ewedihalehu 你这边跑出来的结果是怎样的？

> python main.py --sample_data cifar10

| model | acc|
|-|-|
|ANN accuracy: |0.8973999999999998|
|GNN accuracy: |0.8970999999999998|
|FNN accuracy: |0.8856000000000003|
|SNN accuracy: |0.8765000000000006|

https://openreview.net/forum?id=XrunSYwoLr 这里的 review 有部分数据，似乎可以对得上

**Comment by Ewedihalehu** [comment]
> > > 因此，这项工作中的逆网络被训练为适合 x > 2 的 （1/x），而 0<x <2 无关紧要
> > 
> > 
> > 请问，你跑出来的精度和论文里相同吗，我怎么跑出来差了好多呀，是没有微调的原因吗
> 
> 复现还是没问题的，我尝试着把它这部分给整到 llama2-7B 模型上做情感分类效果还是不错的

那四个pth文件你是直接用的吗

**Comment by condy0919** [comment]
> 那四个pth文件你是直接用的吗

没有，我看了下具体的输入范围，针对性地再训练了一下，特别是 1/sqrt(x)。

**Comment by Ewedihalehu** [comment]
> 的输入范围，针对性地再训练了一下，特别是 1/sqrt(x)。

训练的代码可以分享给我一份吗

**Comment by condy0919** [comment]
> > 的输入范围，针对性地再训练了一下，特别是 1/sqrt(x)。
> 
> 训练的代码可以分享给我一份吗

``` python
import os
import torch
import torch.nn as nn
from torch.utils.data import TensorDataset, DataLoader
from tqdm.auto import tqdm

class Distilled_EXP(nn.Module):
    def __init__(self, num_neurons=32):
        super().__init__()
        self.approximator = nn.Sequential(nn.Linear(1, num_neurons), nn.ReLU(), nn.Linear(num_neurons, 1))

    def forward(self, x):
        return torch.squeeze(self.approximator(torch.unsqueeze(x, -1)))


class Distilled_INV(nn.Module):
    def __init__(self, num_neurons=32):
        super().__init__()
        self.approximator = nn.Sequential(nn.Linear(1, num_neurons), nn.ReLU(), nn.Linear(num_neurons, 1))

    def forward(self, x):
        return torch.squeeze(self.approximator(torch.unsqueeze(x, -1)), dim=-1)


class Distilled_GELU(nn.Module):
    def __init__(self, num_neurons=64):
        super().__init__()
        self.approximator = nn.Sequential(nn.Linear(1, num_neurons), nn.ReLU(), nn.Linear(num_neurons, 1))

    def forward(self, x):
        return torch.squeeze(self.approximator(torch.unsqueeze(x, -1)))


class Distilled_SILU(nn.Module):
    def __init__(self, num_neurons=64):
        super().__init__()
        self.approximator = nn.Sequential(nn.Linear(1, num_neurons), nn.ReLU(), nn.Linear(num_neurons, 1))

    def forward(self, x):
        return torch.squeeze(self.approximator(torch.unsqueeze(x, -1)))


class Distilled_SQRTINV(nn.Module):
    def __init__(self, num_neurons=8):
        super().__init__()
        self.approximator = nn.Sequential(
            nn.Linear(1, num_neurons),
            nn.ReLU(),
            nn.Linear(num_neurons, 1),
        )

    def forward(self, x):
        return torch.squeeze(self.approximator(torch.unsqueeze(x, dim=-1)), dim=-1)


def train_premodels(model, name, num_epochs, lr, device, path):
    # criterion = nn.HuberLoss()
    # optimizer = torch.optim.SGD(model.parameters(), lr=lr)
    criterion = nn.MSELoss()
    optimizer = torch.optim.Adam(model.parameters(), lr=lr)
    scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=num_epochs)

    model.to(device)

    N = 64
    x_train = torch.cat(
        [
            0.995 * torch.rand(N * 82192, device=device) + 0.005,
            2499 * torch.rand(N * 82192, device=device) + 1,
        ],
        dim=0,
    )
    y_train = torch.rsqrt(x_train)
    # y_train = torch.nn.SiLU()(x_train)
    train_dataset = TensorDataset(x_train, y_train)

    train_loader = DataLoader(dataset=train_dataset, batch_size=128, shuffle=True)

    for epoch in range(num_epochs):
        pbar = tqdm(train_loader, desc=f"Epoch {epoch+1}/{num_epochs}", leave=True, ncols=150)

        iter = 0
        for inputs, targets in pbar:
            outputs = model(inputs)
            loss = criterion(outputs, targets)

            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

            with torch.no_grad():
                mse = nn.MSELoss()(outputs, targets)
                acc = 1 - mse.item()

            iter += 1
            if iter % 200 == 0:
                pbar.set_postfix({"Loss": f"{loss.item():.4f}", "Accuracy": f"{acc:.4f}"})

            scheduler.step()
        checkpoint = {
            "epoch": epoch,
            "model_state_dict": model.state_dict(),
            "optimizer_state_dict": optimizer.state_dict(),
        }
        torch.save(checkpoint, f"{path}{name}.pt")


num_neurons = 16
output_path = f"./premodels/ours/"
os.makedirs(output_path, exist_ok=True)
device = "cuda:1"
name = f"our_sqrtinv_{num_neurons}"

model_silu = Distilled_SQRTINV(num_neurons=num_neurons)

train_premodels(model=model_silu, name=name, num_epochs=100, lr=0.01, device=device, path=output_path)
```

**Comment by condy0919** [comment]
问个问题哈，我发现在 spike_x2x 后，获得的矩阵不是 01 矩阵，而是一个 fp16 的矩阵，只是它的值域限制在了 [0, self.burst_T] 范围内，这好像与论文里说的 binary spike 有点对不上吧？ @LockFaiz @Ewedihalehu

**Comment by Ewedihalehu** [comment]
> > > 的输入范围，针对性地再训练了一下，特别是 1/sqrt(x)。
> > 
> > 
> > 训练的代码可以分享给我一份吗
> 
> ```python
> import os
> import torch
> import torch.nn as nn
> from torch.utils.data import TensorDataset, DataLoader
> from tqdm.auto import tqdm
> 
> class Distilled_EXP(nn.Module):
>     def __init__(self, num_neurons=32):
>         super().__init__()
>         self.approximator = nn.Sequential(nn.Linear(1, num_neurons), nn.ReLU(), nn.Linear(num_neurons, 1))
> 
>     def forward(self, x):
>         return torch.squeeze(self.approximator(torch.unsqueeze(x, -1)))
> 
> 
> class Distilled_INV(nn.Module):
>     def __init__(self, num_neurons=32):
>         super().__init__()
>         self.approximator = nn.Sequential(nn.Linear(1, num_neurons), nn.ReLU(), nn.Linear(num_neurons, 1))
> 
>     def forward(self, x):
>         return torch.squeeze(self.approximator(torch.unsqueeze(x, -1)), dim=-1)
> 
> 
> class Distilled_GELU(nn.Module):
>     def __init__(self, num_neurons=64):
>         super().__init__()
>         self.approximator = nn.Sequential(nn.Linear(1, num_neurons), nn.ReLU(), nn.Linear(num_neurons, 1))
> 
>     def forward(self, x):
>         return torch.squeeze(self.approximator(torch.unsqueeze(x, -1)))
> 
> 
> class Distilled_SILU(nn.Module):
>     def __init__(self, num_neurons=64):
>         super().__init__()
>         self.approximator = nn.Sequential(nn.Linear(1, num_neurons), nn.ReLU(), nn.Linear(num_neurons, 1))
> 
>     def forward(self, x):
>         return torch.squeeze(self.approximator(torch.unsqueeze(x, -1)))
> 
> 
> class Distilled_SQRTINV(nn.Module):
>     def __init__(self, num_neurons=8):
>         super().__init__()
>         self.approximator = nn.Sequential(
>             nn.Linear(1, num_neurons),
>             nn.ReLU(),
>             nn.Linear(num_neurons, 1),
>         )
> 
>     def forward(self, x):
>         return torch.squeeze(self.approximator(torch.unsqueeze(x, dim=-1)), dim=-1)
> 
> 
> def train_premodels(model, name, num_epochs, lr, device, path):
>     # criterion = nn.HuberLoss()
>     # optimizer = torch.optim.SGD(model.parameters(), lr=lr)
>     criterion = nn.MSELoss()
>     optimizer = torch.optim.Adam(model.parameters(), lr=lr)
>     scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=num_epochs)
> 
>     model.to(device)
> 
>     N = 64
>     x_train = torch.cat(
>         [
>             0.995 * torch.rand(N * 82192, device=device) + 0.005,
>             2499 * torch.rand(N * 82192, device=device) + 1,
>         ],
>         dim=0,
>     )
>     y_train = torch.rsqrt(x_train)
>     # y_train = torch.nn.SiLU()(x_train)
>     train_dataset = TensorDataset(x_train, y_train)
> 
>     train_loader = DataLoader(dataset=train_dataset, batch_size=128, shuffle=True)
> 
>     for epoch in range(num_epochs):
>         pbar = tqdm(train_loader, desc=f"Epoch {epoch+1}/{num_epochs}", leave=True, ncols=150)
> 
>         iter = 0
>         for inputs, targets in pbar:
>             outputs = model(inputs)
>             loss = criterion(outputs, targets)
> 
>             optimizer.zero_grad()
>             loss.backward()
>             optimizer.step()
> 
>             with torch.no_grad():
>                 mse = nn.MSELoss()(outputs, targets)
>                 acc = 1 - mse.item()
> 
>             iter += 1
>             if iter % 200 == 0:
>                 pbar.set_postfix({"Loss": f"{loss.item():.4f}", "Accuracy": f"{acc:.4f}"})
> 
>             scheduler.step()
>         checkpoint = {
>             "epoch": epoch,
>             "model_state_dict": model.state_dict(),
>             "optimizer_state_dict": optimizer.state_dict(),
>         }
>         torch.save(checkpoint, f"{path}{name}.pt")
> 
> 
> num_neurons = 16
> output_path = f"./premodels/ours/"
> os.makedirs(output_path, exist_ok=True)
> device = "cuda:1"
> name = f"our_sqrtinv_{num_neurons}"
> 
> model_silu = Distilled_SQRTINV(num_neurons=num_neurons)
> 
> train_premodels(model=model_silu, name=name, num_epochs=100, lr=0.01, device=device, path=output_path)
> ```

谢谢！

**Comment by Ewedihalehu** [comment]
> 问个问题哈，我发现在 spike_x2x 后，获得的矩阵不是 01 矩阵，而是一个 fp16 的矩阵，只是它的值域限制在了 [0, self.burst_T] 范围内，这好像与论文里说的 binary spike 有点对不上吧？ @LockFaiz @Ewedihalehu

我传的也不是01，这个问题我也不清楚

**Comment by Ewedihalehu** [comment]
> > > 因此，这项工作中的逆网络被训练为适合 x > 2 的 （1/x），而 0<x <2 无关紧要
> > 
> > 
> > 请问，你跑出来的精度和论文里相同吗，我怎么跑出来差了好多呀，是没有微调的原因吗
> 
> 复现还是没问题的，我尝试着把它这部分给整到 llama2-7B 模型上做情感分类效果还是不错的

请问你的num_neurons都取的多少呢

**Comment by condy0919** [comment]
> > > > 因此，这项工作中的逆网络被训练为适合 x > 2 的 （1/x），而 0<x <2 无关紧要
> > > 
> > > 
> > > 请问，你跑出来的精度和论文里相同吗，我怎么跑出来差了好多呀，是没有微调的原因吗
> > 
> > 
> > 复现还是没问题的，我尝试着把它这部分给整到 llama2-7B 模型上做情感分类效果还是不错的
> 
> 请问你的num_neurons都取的多少呢

与论文里保持一致的。如果你是说在 llama2-7b 上的效果的话，silu 是 64，与 gelu 保持一致。

**Comment by Ewedihalehu** [comment]
> > > > > 因此，这项工作中的逆网络被训练为适合 x > 2 的 （1/x），而 0<x <2 无关紧要
> > > > 
> > > > 
> > > > 请问，你跑出来的精度和论文里相同吗，我怎么跑出来差了好多呀，是没有微调的原因吗
> > > 
> > > 
> > > 复现还是没问题的，我尝试着把它这部分给整到 llama2-7B 模型上做情感分类效果还是不错的
> > 
> > 
> > 请问你的num_neurons都取的多少呢
> 
> 与论文里保持一致的。如果你是说在 llama2-7b 上的效果的话，silu 是 64，与 gelu 保持一致。

好滴谢谢~ 还想请问，你训练的exp silu inv sqr的x_train都取的什么范围呀，我的精度一直上不来，我怀疑是我选的x_train范围不太正确

**Comment by condy0919** [comment]
> 好滴谢谢~ 还想请问，你训练的exp silu inv sqr的x_train都取的什么范围呀，我的精度一直上不来，我怀疑是我选的x_train范围不太正确

这个你可以自己实际跑一下模型，然后sample一下。

**Comment by Ewedihalehu** [comment]
> > 好滴谢谢~ 还想请问，你训练的exp silu inv sqr的x_train都取的什么范围呀，我的精度一直上不来，我怀疑是我选的x_train范围不太正确
> 
> 这个你可以自己实际跑一下模型，然后sample一下。

嗯嗯好滴 你有没有换其他模型结构用这个方法呀

**Comment by condy0919** [comment]
> 嗯嗯好滴 你有没有换其他模型结构用这个方法呀

没有，我只在 llama2-7b 上测试了一下

**Comment by chaibosong** [comment]
> > > > > 因此，这项工作中的逆网络被训练为适合 x > 2 的 （1/x），而 0<x <2 无关紧要
> > > > 
> > > > 
> > > > 请问，你跑出来的精度和论文里相同吗，我怎么跑出来差了好多呀，是没有微调的原因吗
> > > 
> > > 
> > > 复现还是没问题的，我尝试着把它这部分给整到 llama2-7B 模型上做情感分类效果还是不错的
> > 
> > 
> > 请问你的num_neurons都取的多少呢
> 
> 与论文里保持一致的。如果你是说在 llama2-7b 上的效果的话，silu 是 64，与 gelu 保持一致。

你好，麻烦可以分享一下在llama2-7b的训练代码吗，非常感谢

**Comment by condy0919** [comment]
> 你好，麻烦可以分享一下在llama2-7b的训练代码吗，非常感谢

模型修改量有点大，建议还是自己实现哈...

**Comment by Polarisamoon** [comment]
怎么在cifar10上测试都要5个小时呀？我是单卡3090，这正常吗？
`python main.py --sample_data cifar10 `
